File: apps/minecraft/velocity/src/main/java/crabcraft/net/crabUtilities/velocity/api/PublicChatFeed.java
package crabcraft.net.crabUtilities.velocity.api;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/** Thread-safe, in-process replay buffer and subscriber fan-out. */
final class PublicChatFeed implements AutoCloseable {

    private final int recentCapacity;
    private final int subscriberCapacity;
    private final Object stateLock = new Object();
    private final Deque<PublicChatEvent> recent = new ArrayDeque<>();
    private final Set<PublicChatSubscription> subscriptions = new LinkedHashSet<>();
    private boolean closed;

    PublicChatFeed(int recentCapacity, int subscriberCapacity) {
        if (recentCapacity < 1) {
            throw new IllegalArgumentException("recentCapacity must be positive");
        }
        if (subscriberCapacity < 1) {
            throw new IllegalArgumentException("subscriberCapacity must be positive");
        }
        this.recentCapacity = recentCapacity;
        this.subscriberCapacity = subscriberCapacity;
    }

    PublicChatSubscription subscribe(String lastEventId, int defaultReplay) {
        if (defaultReplay < 0) {
            throw new IllegalArgumentException("defaultReplay must not be negative");
        }
        synchronized (stateLock) {
            if (closed) throw new IllegalStateException("public chat feed is closed");

            PublicChatSubscription subscription =
                    new PublicChatSubscription(this, subscriberCapacity);
            for (PublicChatEvent event : replay(lastEventId, defaultReplay)) {
                subscription.offer(event);
            }
            subscriptions.add(subscription);
            return subscription;
        }
    }

    void publish(PublicChatEvent event) {
        synchronized (stateLock) {
            if (closed) return;

            PublicChatEvent newest = recent.peekLast();
            if (newest != null && PublicChatEvent.compareIds(event.id(), newest.id()) <= 0) {
                return;
            }

            recent.addLast(event);
            while (recent.size() > recentCapacity) {
                recent.removeFirst();
            }
            for (PublicChatSubscription subscription : subscriptions) {
                subscription.offer(event);
            }
        }
    }

    int recentCapacity() {
        return recentCapacity;
    }

    int subscriptionCount() {
        synchronized (stateLock) {
            return subscriptions.size();
        }
    }

    void unsubscribe(PublicChatSubscription subscription) {
        synchronized (stateLock) {
            subscriptions.remove(subscription);
        }
    }

    private List<PublicChatEvent> replay(String lastEventId, int defaultReplay) {
        List<PublicChatEvent> snapshot = new ArrayList<>(recent);
        if (PublicChatEvent.isStreamId(lastEventId)) {
            snapshot.removeIf(event -> PublicChatEvent.compareIds(event.id(), lastEventId) <= 0);
            return snapshot;
        }

        int count = Math.min(defaultReplay, snapshot.size());
        return snapshot.subList(snapshot.size() - count, snapshot.size());
    }

    @Override
    public void close() {
        List<PublicChatSubscription> toClose;
        synchronized (stateLock) {
            if (closed) return;
            closed = true;
            toClose = new ArrayList<>(subscriptions);
            subscriptions.clear();
            recent.clear();
        }
        for (PublicChatSubscription subscription : toClose) {
            subscription.closeFromFeed();
        }
    }
}

File: apps/minecraft/velocity/src/main/kotlin/crabcraft/net/crabUtilities/velocity/api/PublicChatFeed.kt
package crabcraft.net.crabUtilities.velocity.api

import java.util.ArrayDeque

/** Thread-safe, in-process replay buffer and subscriber fan-out. */
class PublicChatFeed(private val recentCapacity: Int, private val subscriberCapacity: Int) : AutoCloseable {
    private val stateLock = Any()
    private val recent = ArrayDeque<PublicChatEvent>()
    private val subscriptions = LinkedHashSet<PublicChatSubscription>()
    private var closed = false

    init {
        require(recentCapacity >= 1) { "recentCapacity must be positive" }
        require(subscriberCapacity >= 1) { "subscriberCapacity must be positive" }
    }

    fun subscribe(lastEventId: String?, defaultReplay: Int): PublicChatSubscription {
        require(defaultReplay >= 0) { "defaultReplay must not be negative" }
        synchronized(stateLock) {
            check(!closed) { "public chat feed is closed" }
            val subscription = PublicChatSubscription(this, subscriberCapacity)
            for (event in replay(lastEventId, defaultReplay)) subscription.offer(event)
            subscriptions.add(subscription)
            return subscription
        }
    }

    fun publish(event: PublicChatEvent) {
        synchronized(stateLock) {
            if (closed) return
            val newest = recent.peekLast()
            if (newest != null && PublicChatEvent.compareIds(event.id(), newest.id()) <= 0) return
            recent.addLast(event)
            while (recent.size > recentCapacity) recent.removeFirst()
            for (subscription in subscriptions) subscription.offer(event)
        }
    }

    fun recentCapacity() = recentCapacity

    fun subscriptionCount(): Int = synchronized(stateLock) { subscriptions.size }

    fun unsubscribe(subscription: PublicChatSubscription) {
        synchronized(stateLock) { subscriptions.remove(subscription) }
    }

    private fun replay(lastEventId: String?, defaultReplay: Int): List<PublicChatEvent> {
        val snapshot = ArrayList(recent)
        if (PublicChatEvent.isStreamId(lastEventId)) {
            snapshot.removeIf { PublicChatEvent.compareIds(it.id(), lastEventId!!) <= 0 }
            return snapshot
        }
        val count = minOf(defaultReplay, snapshot.size)
        return snapshot.subList(snapshot.size - count, snapshot.size)
    }

    override fun close() {
        val toClose =
            synchronized(stateLock) {
                if (closed) return
                closed = true
                val snapshot = ArrayList(subscriptions)
                subscriptions.clear()
                recent.clear()
                snapshot
            }
        for (subscription in toClose) subscription.closeFromFeed()
    }
}
